### ISSUE TITLE
Generated Service Definition types are inaccurate

### ISSUE THREAD

**Issue by mintchkin** [grpc/grpc-node#1740]
### Problem description

The typescript generator introduced in #1474 produces an incorrect/incomplete type for services in loaded protos. Referencing the [golden-generated output](https://github.com/grpc/grpc-node/blob/5d405178f3f9834cd52565f9f07a22d0a34ea8a9/packages/proto-loader/golden-generated/echo.ts#L81), the type of the `service` field is [`ServiceDefinition`](https://github.com/grpc/grpc-node/blob/5d405178f3f9834cd52565f9f07a22d0a34ea8a9/packages/proto-loader/src/index.ts#L131-L133).

This causes problems with the type of `server.addService` in [the native grpc package](https://github.com/grpc/grpc-node/blob/545a9359a684304dddcf93270a6c1add82d04477/packages/grpc-native-core/index.d.ts#L266-L269):

```ts
const packageDefinition = (protoLoader.loadSync(...) as unknown) as ProtoGrpcType;
const server = new grpc.Server();
server.addService(
    packageDefinition.google.showcase.v1beta1.Echo.service,
    // ~~~~~~ TS: Argument of type 'ServiceDefinition' is not assignable to parameter of type 'ServiceDefinition<{ Block: ..., Chat: ..., ...}>'
    { Block: ..., Chat: ..., ... }
)
```

I think the type can be improved without changing the generated code much by injecting a custom `ServiceDefinition` type based on the one in [the native grpc package](https://github.com/grpc/grpc-node/blob/545a9359a684304dddcf93270a6c1add82d04477/packages/grpc-native-core/index.d.ts#L315-L317), or maybe the [related one in `@grpc/proto-loader`](https://github.com/grpc/grpc-node/blob/5d405178f3f9834cd52565f9f07a22d0a34ea8a9/packages/proto-loader/src/index.ts#L131-L133) could just be replaced instead.

```ts
type ServiceDefinition<
  ImplementationType extends grpc.UntypedServiceImplementation = grpc.UntypedServiceImplementation
> = {
  [K in keyof ImplementationType]: MethodDefinition<
    Parameters<ImplementationType[K]>[0], // Request Type
    ReturnType<ImplementationType[K]> // Return Type
  >;
};

// ... later in e.g. echo.ts

Echo: SubtypeConstructor<typeof grpc.Client, _google_showcase_v1beta1_EchoClient> & { service: ServiceDefinition<EchoHandlers> }
```

### Environment
 - `@grpc/proto-loader@0.6.0`

**Comment by murgatroid99** [comment]
One solution here may be to loosen the type parameter in the native grpc package to avoid this kind of problem. That current type may not even be correct in all cases.

Another good alternative may be to explicitly generate the service definition type along with each service implementation type. In your suggestion you are using generics to extract the service definition type from the implementation type, but we can take advantage of the fact that the details of the implementation type are known at generation time.

So, it would be something like this in `google/showcase/v1beta1/Echo.ts`:

```ts
export interface EchoServiceDefinition {
  Block: MethodDefinition<_google_showcase_v1beta1_BlockRequest__Output, _google_showcase_v1beta1_BlockResponse>,
  // etc.
}
```

It would even be possible to go a step further and generate the exact contents of each method definition as a type.

**Comment by murgatroid99** [comment]
This has been fixed in version 0.6.1, with types similar to the example code I put in my previous comment.
